IdentityCore/src/util/mac/MSIDKeychainUtil.m

```diff
@@ -22,48 +22,71 @@
 // THE SOFTWARE.
 
 #import "MSIDKeychainUtil.h"
+#import "MSIDKeychainUtil+Internal.h"
 
 @implementation MSIDKeychainUtil
 
-+ (NSString *)teamId
+- (instancetype)init
 {
-    static dispatch_once_t once;
-    static NSString *keychainTeamId = nil;
+    return [[self class] sharedInstance];
+}
+
+- (instancetype)initPrivate
+{
+    self = [super init];
+    if (self)
+    {
+        self.teamId = [self getTeamId];
+    }
+    
+    return self;
+}
+
++ (MSIDKeychainUtil *)sharedInstance
+{
+    static MSIDKeychainUtil *singleton = nil;
+    static dispatch_once_t onceToken;
     
-    dispatch_once(&once, ^{
-        SecCodeRef selfCode = NULL;
-        SecCodeCopySelf(kSecCSDefaultFlags, &selfCode);
-        
-        if (selfCode)
-        {
-            CFDictionaryRef cfDic = NULL;
-            SecCodeCopySigningInformation(selfCode, kSecCSSigningInformation, &cfDic);
-            NSDictionary* signingDic = CFBridgingRelease(cfDic);
-            keychainTeamId = [signingDic objectForKey:(__bridge NSString*)kSecCodeInfoTeamIdentifier];
-            
-            MSID_LOG_NO_PII(MSIDLogLevelInfo, nil, nil, @"Using \"%@\" Team ID.", _PII_NULLIFY(keychainTeamId));
-            MSID_LOG_PII(MSIDLogLevelInfo, nil, nil, @"Using \"%@\" Team ID.", keychainTeamId);
-            
-            CFRelease(selfCode);
-        }
+    dispatch_once(&onceToken, ^{
+        singleton = [[self alloc] initPrivate];
     });
     
+    return singleton;
+}
+
+- (NSString *)getTeamId
+{
+    NSString *keychainTeamId = nil;
+    SecCodeRef selfCode = NULL;
+    SecCodeCopySelf(kSecCSDefaultFlags, &selfCode);
+    
+    if (selfCode)
+    {
+        CFDictionaryRef cfDic = NULL;
+        SecCodeCopySigningInformation(selfCode, kSecCSSigningInformation, &cfDic);
+        NSDictionary* signingDic = CFBridgingRelease(cfDic);
+        keychainTeamId = [signingDic objectForKey:(__bridge NSString*)kSecCodeInfoTeamIdentifier];
```

Comment:
should we log an error if either cfdic or keychainteamId is nil?

Reply:
Done